### ISSUE TITLE
Providing values for the Amax parameters

### ISSUE THREAD

**Issue by agruss2** [jhpoelen/fb-osmose-bridge#23]
The Amax parameter, i.e., the threshold age above which a von Bertalanffy growth equation can be assumed (and below which linear growth is assumed) can be obtained directly from FishBase/SeaLifeBase for some species. For those species for which the Amax parameter cannot be obtained directly from FishBase/SeaLifeBase, the Amax parameter can be obtained from the age at metamorphosis (i.e., the age corresponding to the size at metamorphosis Lx of the marine organism under consideration). 
Question to Deng and Skit: Are Lx estimates provided in FishBase/SeaLifeBase?

**Comment by agruss2** [comment]
@FIN-JBarile @Dengaloo 
We need to make a decision regarding the estimation of the species.vonbertalanffy.threshold.age.sp parameters, i.e., the threshold ages below which a linear growth model is used to calculate mean length increase. 
In general, the species.vonbertalanffy.threshold.age.sp parameter is set to 1 year for all the focal functional groups represented in a given OSMOSE model. An exception to this general pattern occurs for focal functional groups with a very short lifespan. For example, in the OSMOSE model of the Southern Benguela, the species.vonbertalanffy.threshold.age.sp parameter of euphausiids was set to 0.17 years, while the species.vonbertalanffy.threshold.age.sp parameter of all the other focal functional groups represented in that OSMOSE model was set to 1 year. 
The OSMOSE model of the West Florida Shelf ("OSMOSE-WFS") differs from the other OSMOSE applications in that the species.vonbertalanffy.threshold.age.sp parameter of each of the focal functional groups represented in OSMOSE-WFS was determined based on the study from which von Bertalanffy growth parameters were taken. More precisely, the species.vonbertalanffy.threshold.age.sp parameter of each of the focal functional groups represented in OSMOSE-WFS was determined based on the smallest individuals in the dataset that allowed to estimate a von Bertalanffy growth function for the focal functional group.   
I do not think that it is a good idea to consider that the species.vonbertalanffy.threshold.age.sp parameter can be equated to the age at metamorphosis (i.e., the age corresponding to the size at metamorphosis Lx of the marine organism under consideration). If we did that, we would significantly underestimate the threshold age above which a von Bertalanffy growth equation can be assumed (and below which linear growth is assumed) for focal functional groups. 
I noticed that in FishBase's poplw.csv file, there is a column called "LengthMin", which is the  	length of the smallest fish considered for estimating a length-weight relationship. Unfortunately, there is no similar column in FishBase's popgrowth.csv file. 
Therefore, my questions for Deng and Skit are: Would it be reasonable to equate the species.vonbertalanffy.threshold.age.sp parameter of a functional group represented in an OSMOSE model to its "LengthMin" in FishBase's poplw.csv file? If not, how do you think that we could obtain a reasonable value from FishBase for the species.vonbertalanffy.threshold.age.sp parameter of each of the functionals group represented in an OSMOSE model?

**Comment by Dengaloo** [comment]
@agruss2 again, not all species will have VBGF growth parameters, in such cases, we use Lmax. Would it be acceptable to use Lmax as Linf and use an estimate K from theta prime for similar species (in length in temperature preferred?)?

**Comment by agruss2** [comment]
@Dengaloo **The parameter that we are considering here is an age estimate**, and not a length estimate. 
I think that, here as well, we should keep things simple. 
Here is how I envision the estimation of the “species.vonbertalanffy.threshold.age.sp” parameter:
(1) We assume that the "LengthMin" parameter provided in FishBase's poplw.csv file is the minimum length at which the species under consideration were measured, and we estimate the “threshold age” parameter from the value of "LengthMin" using the inverse von Bertalanffy growth equation for the species under consideration. 
(2) If, for the species under consideration, we do not have any "LengthMin" parameter in FishBase's poplw.csv file or we do not have any von Bertalanffy growth equation (and, therefore, we do not have any inverse von Bertalanffy growth equation), then we assume that the “threshold age” for this species is equal to 1 year, i.e., we set the “species.vonbertalanffy.threshold.age.sp” parameter of this species to 1. 
Do you like these ideas?

**Comment by Dengaloo** [comment]
@agruss2 : OK (1) is doable and (2) is practical, I like it. We should be able to see which species have lifespans of 1 year or less, and if these will actually be important enough to merit investigation. I can look it up.

**Comment by agruss2** [comment]
@Dengaloo Actually, I think that **some analyses are needed to derive the rules of thumb**. 

Could you please? 
**(1)** Determine all the species in FishBase and SeaLifeBase for which you have estimates for both the "LengthMin" and "LengthMax" parameters (from the "POPLW" tab; 
**(2)** For the selected species, estimate the age corresponding to "LengthMin" (let's call this "AgeMin") and the age corresponding to "LengthMax" (let's call this "AgeMax"), using the inverse von Bertalanffy growth equation (of course, for those species for which it is possible to use a von Bertalanffy growth equation); 
**(3)** For the species selected in previous step, estimate the ratio of AgeMin over AgeMax (AgeMin/AgeMax). 
**(4)** Calculate a mean AgeMin/AgeMax, using all the estimates of AgeMin/AgeMax that you have; this mean AgeMin/AgeMax will be used by the bridge to derive an estimate of "threshold age" for those species for which there is no estimate of "LengthMin" in the "POPLW" tab, as follows: 
threshold age = (Maximum age of the species) * (AgeMin/AgeMax)

Does all this sound good to you?

**Comment by Dengaloo** [comment]
@agruss2 : there are 408 nonfishes and 4276 fishes with both LengthMin and LengthMax data in SLB and FB, respectively. so yes, we can test this. Give me a day to do this.

**Comment by agruss2** [comment]
@Dengaloo Sounds excellent; thanks!

**Comment by Dengaloo** [comment]
@agruss2 : I have a problem actually, the POPLW length min and length max parameters refer to the minimum and maximum sizes of the sample used for the length weight analysis. Often, these lengths are representative of the population, but in some cases, these are very limited to a particular size range that is the focus of the study. What I will do is eliminate those instances where the size range of the sample is too narrow or too near to Lmax (so I will have less samples, but probably still viable for a model). Do you agree?

**Comment by agruss2** [comment]
@Dengaloo You raised a very good point; and I agree with your plan!

**Comment by agruss2** [comment]
@Dengaloo Please let me know when you are done with this. Then, I will be able to write a new issue for Jorrit, so that the API implements what is described above. Many thanks !

**Comment by Dengaloo** [comment]
@agruss2 : trying to leverage data that we have in FB and SLB, working with research assistants here to clean up the data so we have a viable data set for the parameter set up. I see that you moved the milestone, thanks for this. will get back to you when I have a working model.

**Comment by agruss2** [comment]
@Dengaloo Sounds great, thanks !

**Comment by Dengaloo** [comment]
@agruss2 : as a simple solution to Age maximum, please use length maximum at 95% of Linfinity; then calculate age for Lmax as t0 + 3/K; t0 = 0 where it is not given. However, you can calculate t0 from Linfinity and K using: log (-t0) = -0.3922 - 0.2752 log Linf - 1.038 log K.

If there is a measure for Lmax in FishBase, but for which no growth K is available, you can calculate K from Lmax using: ThetaPrime=logK + 2*logLinf for all species in a genus and/or family. I will not go as far as order and class with this if you want to keep some detail as to the growth.

If you can get Jorrit to write an algorithm that implements these rules, then we have Amax fixed.

**Comment by Dengaloo** [comment]
@agruss2 : I think we should stick to these algorithms because we do not want to repeat the process again with different data sets. In any case, I don't think that there will be much of a difference.

I forgot to mention that if we used ThetaPrime, we will basically be using the principle of the auximetric grid, but using lengths instead of weights.

ThetaPrime works for fish, but will not work for standardizing the tmax for other animal groups. If we want to standardize tmax to include invertebrates as well, we will need to use Theta, i.e., auximetric grid using weights.

Let me know if you want me to design algorithms for Theta instead.

**Comment by Dengaloo** [comment]
@agruss2 and @FIN-JBarile 

I have now developed the empirical model for estimating maximum weight for all fishes, other vertebrates and invertebrates, with a correction for jellyfishes which are 95% water;

the equation using n=5367 pairs of Woo and K values, an r2=0.3336 is
logK=-0.2372logWoo+0.2282
establishes the auximetric relationship in weight

i will need to write a paper about this important relationship because this now allows us to compare growth across taxonomic groups

given this, we can establish maximum weights at 0.86 of Woo, and estimate maximum age from maximum weight given the von Bertalanffy weight growth equation of Wt=Woo*(1-e^-Kt)^3, assuming that t0=0

@FIN-JBarile might want to add the Woo and K values for all marine animals both fish and nonfishes, in the Estimates table, and providing the Amax (from Woo) as another field. Amax from Woo would then be an estimate of the longevity of the population.

**Comment by agruss2** [comment]
@Dengaloo I am very sorry, but this is not the parameter that we need for OSMOSE. I am very sorry, you went to a wrong direction for this specific issue, because the name of the parameter considered in the present issue ("Amax") is admittedly confusing ...
**The "Amax" parameter in OSMOSE** is the threshold age above which a von Bertalanffy growth equation can be assumed (and, therefore, below which linear growth is assumed). This is not an equivalent of longevity. 
Please let me know if I missed something here, and how you think that we could get estimates of the threshold age above which a von Bertalanffy growth equation can be assumed for the different fish and invertebrate species represented in OSMOSE

**Comment by Dengaloo** [comment]
@agruss2 : so is Amax related to length at first maturity?

**Comment by Dengaloo** [comment]
@agruss2 sorry but I am now totally confused. I copy your previous comment here:

(1) Determine all the species in FishBase and SeaLifeBase for which you have estimates for both the "LengthMin" and "LengthMax" parameters (from the "POPLW" tab;
(2) For the selected species, estimate the age corresponding to "LengthMin" (let's call this "AgeMin") and the age corresponding to "LengthMax" (let's call this "AgeMax"), using the inverse von Bertalanffy growth equation (of course, for those species for which it is possible to use a von Bertalanffy growth equation);
(3) For the species selected in previous step, estimate the ratio of AgeMin over AgeMax (AgeMin/AgeMax).
(4) Calculate a mean AgeMin/AgeMax, using all the estimates of AgeMin/AgeMax that you have; this mean AgeMin/AgeMax will be used by the bridge to derive an estimate of "threshold age" for those species for which there is no estimate of "LengthMin" in the "POPLW" tab, as follows:
threshold age = (Maximum age of the species) * (AgeMin/AgeMax)

So you do want to have an estimate of longevity. You want to have observed AgeMin and AgeMax as you describe in (4).

But, you want to use the POPLW table for this. The POPLW table as I have already mentioned is not a good source because the LengthMin and LengthMax in that table only refer to the sample used in the POPLW study. Nothing to do with observed min or max lengths.

So, I am now at a loss as to how to proceed. I will let you try to clarify this before I do anything else.

Thanks

**Comment by agruss2** [comment]
@Dengaloo Sorry about that, I am the one who introduced confusion here !!
So, the threshold age is the Amax parameter. 
And AgeMax is the age corresponding to age corresponding to "LengthMax". 
Very sorry for having introduced confusion here! Please let me rephrase everythign below.

**Comment by agruss2** [comment]
@Dengaloo So, all that remains to be done here is to have a means to estimate the "threshold age" for any species (both fishes and nonfishes):
threshold age = (Maximum age of the species) * (AgeMin/AgeMax)

-> You have established a way to calculate AgeMax for any species (both fishes and nonfishes; detailed above) and have asked @FIN-JBarile to add the Woo and K values for both fish and nonfishes, in the ESTIMATE table as well as to provide AgeMax (from Woo) as another field.
Hence, we seem to be all good regarding the estimation of the AgeMax parameter and as to where the API can retrieve AgeMax estimates. 

-> Now, regarding AgeMin: How can we calculate this parameter for any species (both fishes and nonfishes)? And where could we store AgeMin estimates so that they are easily retrieved by the API? 

Many thanks!

**Comment by FIN-JBarile** [comment]
@Dengaloo I need specifications for adding Woo (Winfinity) and K values to the ESTIMATE table. We can take this up separately.

**Comment by agruss2** [comment]
@Dengaloo @FIN-JBarile  I am extending the deadline of this issue to March, 12th. Many thanks!

**Comment by agruss2** [comment]
@Dengaloo @FIN-JBarile  There have not been any discussions on the estimation of parameters for OSMOSE during the past weeks. Hopefully, these discussions will resume next week, so that I can open new issues for Jorrit to estimate values for the remaining OSMOSE parameters. I am extending the deadline of this issue again. Thanks a lot.

**Comment by Dengaloo** [comment]
@agruss2 @FIN-JBarile : OK so we are in the hunt for AgeMin. If I understand things correctly, we want to know when the species starts to make an impact/contribution to the ecosystem. So, if we assume that this happens at the first instance that the larval stage turns into the juvenile stage, I think we can assume this as age 0.5, before the species turns 1 year old. For most fish species, this will work. Didn't we say that this what we were going to do? In any case, you also might want to consider pegging this AgeMin with age at first maturity - because this would be when feeding would be most important, i.e., immature individuals tend to feed for growth rather than for reproduction. If this is a good description of AgeMin, then representing AgeMin=50% Age at first maturity, would be doable.

**Comment by agruss2** [comment]
@Dengaloo @FIN-JBarile AgeMin is not the age where the species starts making a contribution to the ecosystem.
AgeMin is the age where we can assume that the species starts growing according to a von Bertalanffy growth equation. 
Then, I would think that equating AgeMin to the age where the species transitions from the larval to the juvenile stage would largely underestimate AgeMin. 
The other extreme would be to equate AgeMin to the age at first maturity, which would lead us to largely overestimate AgeMin. 
Equating AgeMin to 50% of the age at first maturity may work. However, what is the rationale here? If we make this assumption, we need a very good rationale. I understand the point you made that “if we assume normality in the distribution, then we can say that most of immature individuals would be at the mid-point of this distribution between age 0 and age at first maturity.” However, what, from an ecological standpoint, back up the assumption that a species starts growing according to a von Bertalanffy growth equation when its age reaches half of its age at first maturity??

**Comment by Dengaloo** [comment]
@agruss2 @FIN-JBarile : We have this table in the larvae database where we have the meristic characters of larvae, and we have the min, mod, and max lengths at first feeding of the larvae. Usually, first feeding is when the larvae becomes a fingerling. Will the maximum length at first feeding be a good equivalent to AgeMin?

see: http://www.fishbase.us/LarvalBase/reproduction/LarvaeInfoSummary.php?stockcode=79&genusname=Gadus&speciesname=morhua&LarvalArea=Northeastern%20Atlantic%20(British%20coasts);%20Northwestern%20Atlantic%20(Delaware%20Bay)

scroll down to the bottom page for the meristic characters.

**Comment by Dengaloo** [comment]
@agruss2 @FIN-JBarile : thinking about this again though, larval dynamics are so different from that of the adults. I am out of ideas Arnaud, because I could not find anything to back up the assumption of 50% of age at first maturity.

though i know that from fingerling size onwards, fish would grow according to the VBGF. then we would need to find studies that say so. we do not have this in fishbase because it has not been a parameter that can be used for adult dynamics.

**Comment by Dengaloo** [comment]
@agruss2 : see this article: http://www.journals.uchicago.edu/doi/pdfplus/10.1086/285995; i think this argues for using age at first maturity as the cut-off point for vbgf growth.

**Comment by Dengaloo** [comment]
@agruss2 : and see this article: http://rstb.royalsocietypublishing.org/content/363/1502/2453.short.

**Comment by Dengaloo** [comment]
@agruss2 both articles seem to support the length at first maturity assumption that i suggested

**Comment by agruss2** [comment]
@Dengaloo Many thanks for having reflected on this issue. 
As I said previously, based on my experience wth the "threshold age" parameter, if we equated AgeMin to the age at first maturity, we would largely overestimate AgeMin.
Therefore, let's please equate AgeMin to 50% of the age at first maturity. I had a look at my OSMOSE data, and this assumption is reasonable. Indeed, for most of the OSMOSE species for which I have estimates for both the "threshold age" parameter and age at first maturity, threshold age is equal to or very close to 50% of the age at first maturity. 
Please let me know if you are OK with us equating AgeMin to 50% of the age at first maturity.
Once I have heard from you, I'll be able to close the present issue and to open a new API issue for Jorrit so that the API can provide estimates for the "threshold age" parameters for OSMOSE focal functional groups.
Thanks a lot!

**Comment by Dengaloo** [comment]
@agruss2 : yes Arnaud, 50% of Lm was my first suggestion. If you are happy with that, then let's go with it.

**Comment by Dengaloo** [comment]
@agruss2 @FIN-JBarile : Skit and I spoke about which parameters to provide for this in the OSMOSE tables. Parameters are Linf, K and Lm; the equation to get age at first maturity is: tm = t0 - ln(1 - Lm / Linf) / K. So, you will take 50% of that age at first maturity as the AgeMin estimate.

**Comment by Dengaloo** [comment]
@agruss2 once Skit has provided these columns in the OSMOSE table, please close the issue. Thanks

**Comment by agruss2** [comment]
@Dengaloo I agree, we will assume that AgeMin = tm/2 = (t0 - ln(1 - Lm / Linf) / K)/ 2
@FIN-JBarile Please let me know when AgeMax and AgeMin are both included in the ESTIMATE table. (For the estimation of AgeMax, please see our comments above). Once I have heard from you, I will close the present issue and will create a new API issue for Jorrit. 
Many thanks to both of you.

**Comment by agruss2** [comment]
@FIN-JBarile 
Regarding the following GitHub issue: https://github.com/jhpoelen/fb-osmose-bridge/issues/23 
Could you please add the four following parameters to the ESTIMATE table: Woo, K, AgeMax and AgeMin? 
(@Dengaloo – Please let us if everything below is correct).

(1) AgeMax is the age of the oldest fish/invertebrate considered for estimating a length-weight relationship. 
(i) Deng developed the following empirical model for all fishes, other vertebrates and invertebrates, with a correction for jellyfishes which are 95% water:
logK = -0.2372.logWoo + 0.2282
(ii) From this model, we can estimate Woo and, therefore, maximum weight (Wmax), which is given by:
Wmax = 0.86.Woo
(iii) Finally, it is possible to estimate AgeMax from Wmax given the von Bertalanffy weight growth equation of Wt=Woo*(1-e^-Kt)^3, assuming that t0=0. 

(2) AgeMin is the age of the youngest fish/invertebrate considered for estimating a length-weight relationship.
This parameter is calculated as follows:
AgeMin = (t0 - ln(1 - Lm / Linf) / K)/ 2

Many thanks, 
Arnaud

**Comment by agruss2** [comment]
@FIN-JBarile 
Could you please add the following four parameters to the ESTIMATE table: Woo, K, AgeMax and AgeMin?
(@Dengaloo – Please let us if everything below is correct).

(1) AgeMax is the age of the oldest fish/invertebrate considered for estimating a length-weight relationship.
(i) Deng developed the following empirical model for all fishes, other vertebrates and invertebrates, with a correction for jellyfishes which are 95% water:
logK = -0.2372.logWoo + 0.2282
(ii) From this model, we can estimate Woo and, therefore, maximum weight (Wmax), which is given by:
Wmax = 0.86.Woo
(iii) Finally, it is possible to estimate AgeMax from Wmax given the von Bertalanffy weight growth equation of Wt=Woo*(1-e^-Kt)^3, assuming that t0=0.

(2) AgeMin is the age of the youngest fish/invertebrate considered for estimating a length-weight relationship.
This parameter is calculated as follows:
AgeMin = (t0 - ln(1 - Lm / Linf) / K)/ 2

Many thanks,
Arnaud

**Comment by FIN-JBarile** [comment]
@agruss2 Here are the [ESTIMATE tables ](https://www.dropbox.com/s/1czlzbk7rlbbwwh/ESTIMATE_FBSLB.zip?dl=0) for fish and non-fish, with estimates of AgeMin and AgeMax (2395 for fish and 310 for non-fish), as well as Winf and K . 

Gelli of SeaLifeBase, in consultation with Deng, populated the values using the above rules.

Note: These are the same ESTIMATE tables provided in Issue #32.

cc: @Dengaloo

**Comment by jhpoelen** [comment]
@FIN-JBarile @agruss2 I've created a new patch release for fishbase/sealifebase tables used by our bridge https://github.com/jhpoelen/fishbase_archiver/releases/tag/v0.2.1-patch2 . Newly introduced variables should be available for mapping.

**Comment by agruss2** [comment]
@FIN-JBarile @jhpoelen Thank you very much for all your work on this. I just updated the "fishbase-mapping.csv" file accordingly. I am going to close the present issue and open a new API issue.
